README
NAME
    Catalyst::Model::DBIC::Schema - DBIx::Class::Schema Model Class

SYNOPSIS
    Manual creation of a DBIx::Class::Schema and a
    Catalyst::Model::DBIC::Schema:

    1.  Create the DBIx:Class schema in MyApp/Schema/FilmDB.pm:

          package MyApp::Schema::FilmDB;
          use base qw/DBIx::Class::Schema/;

          __PACKAGE__->load_classes(qw/Actor Role/);

    2.  Create some classes for the tables in the database, for example an
        Actor in MyApp/Schema/FilmDB/Actor.pm:

          package MyApp::Schema::FilmDB::Actor;
          use base qw/DBIx::Class/

          __PACKAGE__->load_components(qw/Core/);
          __PACKAGE__->table('actor');

          ...

        and a Role in MyApp/Schema/Role.pm:

          package MyApp::Schema::FilmDB::Role;
          use base qw/DBIx::Class/

          __PACKAGE__->load_components(qw/Core/);
          __PACKAGE__->table('role');

          ...    

        Notice that the schema is in MyApp::Schema, not in MyApp::Model.
        This way it's usable as a standalone module and you can test/run it
        without Catalyst.

    3.  To expose it to Catalyst as a model, you should create a DBIC Model
        in MyApp/Model/FilmDB.pm:

          package MyApp::Model::FilmDB;
          use base qw/Catalyst::Model::DBIC::Schema/;

          __PACKAGE__->config(
              schema_class => 'MyApp::Schema::FilmDB',
              connect_info => [
                                "DBI:...",
                                "username",
                                "password",
                                {AutoCommit => 1}
                              ]
          );

        See below for a full list of the possible config parameters.

    Now you have a working Model, accessing your separate DBIC Schema. Which
    can be used/accessed in the normal Catalyst manner, via $c->model():

      my $actor = $c->model('FilmDB::Actor')->find(1);

    You can also use it to set up DBIC authentication with
    Authentication::Store::DBIC in MyApp.pm:

      package MyApp;

      use Catalyst qw/... Authentication::Store::DBIC/;

      ...

      __PACKAGE__->config->{authentication}{dbic} = {
          user_class      => 'FilmDB::Actor',
          user_field      => 'name',
          password_field  => 'password'
      }

    "$c->model()" returns a DBIx::Class::ResultSet for the source name
    parameter passed. To find out more about which methods can be called on
    a ResultSet, or how to add your own methods to it, please see the
    ResultSet documentation in the DBIx::Class distribution.

    Some examples are given below:

      # to access schema methods directly:
      $c->model('FilmDB')->schema->source(...);

      # to access the source object, resultset, and class:
      $c->model('FilmDB')->source(...);
      $c->model('FilmDB')->resultset(...);
      $c->model('FilmDB')->class(...);

      # For resultsets, there's an even quicker shortcut:
      $c->model('FilmDB::Actor')
      # is the same as $c->model('FilmDB')->resultset('Actor')

      # To get the composed schema for making new connections:
      my $newconn = $c->model('FilmDB')->composed_schema->connect(...);

      # Or the same thing via a convenience shortcut:
      my $newconn = $c->model('FilmDB')->connect(...);

      # or, if your schema works on different storage drivers:
      my $newconn = $c->model('FilmDB')->composed_schema->clone();
      $newconn->storage_type('::LDAP');
      $newconn->connection(...);

      # and again, a convenience shortcut
      my $newconn = $c->model('FilmDB')->clone();
      $newconn->storage_type('::LDAP');
      $newconn->connection(...);

DESCRIPTION
    This is a Catalyst Model for DBIx::Class::Schema-based Models. See the
    documentation for Catalyst::Helper::Model::DBIC::Schema for information
    on generating these Models via Helper scripts.

CONFIG PARAMETERS
    schema_class
        This is the classname of your DBIx::Class::Schema Schema. It needs
        to be findable in @INC, but it does not need to be inside the
        "Catalyst::Model::" namespace. This parameter is required.

    connect_info
        This is an arrayref of connection parameters, which are specific to
        your "storage_type" (see your storage type documentation for more
        details).

        This is not required if "schema_class" already has connection
        information defined inside itself (which isn't highly recommended,
        but can be done)

        For DBIx::Class::Storage::DBI, which is the only supported
        "storage_type" in DBIx::Class at the time of this writing, the
        parameters are your dsn, username, password, and connect options
        hashref.

        See "connect_info" in DBIx::Class::Storage::DBI for a detailed
        explanation of the arguments supported.

        Examples:

          connect_info => [ 'dbi:Pg:dbname=mypgdb', 'postgres', '' ],

          connect_info => [
                            'dbi:SQLite:dbname=foo.db',
                            {
                              on_connect_do => [
                                'PRAGMA synchronous = OFF',
                              ],
                            }
                          ],

          connect_info => [
                            'dbi:Pg:dbname=mypgdb',
                            'postgres',
                            '',
                            { AutoCommit => 0 },
                            {
                              on_connect_do => [
                                'some SQL statement',
                                'another SQL statement',
                              ],
                            }
                          ],

    storage_type
        Allows the use of a different "storage_type" than what is set in
        your "schema_class" (which in turn defaults to "::DBI" if not set in
        current DBIx::Class). Completely optional, and probably unnecessary
        for most people until other storage backends become available for
        DBIx::Class.

METHODS
    new Instantiates the Model based on the above-documented ->config
        parameters. The only required parameter is "schema_class".
        "connect_info" is required in the case that "schema_class" does not
        already have connection information defined for it.

    schema
        Accessor which returns the connected schema being used by the this
        model. There are direct shortcuts on the model class itself for
        schema->resultset, schema->source, and schema->class.

    composed_schema
        Accessor which returns the composed schema, which has no connection
        info, which was used in constructing the "schema" above. Useful for
        creating new connections based on the same schema/model. There are
        direct shortcuts from the model object for composed_schema->clone
        and composed_schema->connect

    clone
        Shortcut for ->composed_schema->clone

    connect
        Shortcut for ->composed_schema->connect

    source
        Shortcut for ->schema->source

    class
        Shortcut for ->schema->class

    resultset
        Shortcut for ->schema->resultset

    storage
        Provides an accessor for the connected schema's storage object. Used
        often for debugging and controlling transactions.

SEE ALSO
    General Catalyst Stuff:

    Catalyst::Manual, Catalyst::Test, Catalyst::Request, Catalyst::Response,
    Catalyst::Helper, Catalyst,

    Stuff related to DBIC and this Model style:

    DBIx::Class, DBIx::Class::Schema, DBIx::Class::Schema::Loader,
    Catalyst::Helper::Model::DBIC::Schema

AUTHOR
    Brandon L Black, "blblack@gmail.com"

COPYRIGHT
    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Perl extension Catalyst::Model::DBIC::Schema

0.67
        - replace MooseX::MarkAsMethods with namespace::autoclean (RT#100003)

0.66  2023-07-30 07:45:00
        - fix Makefile.PL to work properly in newer perls without . in @INC
        - update repository link in metadata
        - fix version requirement in test module
        - fix link in Pod
        - work around hack in Catalyst-Runtime 5.90080

0.65  2014-08-04 11:00:00
        - Make compose_namespaces and install model shortcuts optional

0.64  2014-07-22 23:00:00
        - Fix Replicated trait that got busted when someone assumed
          Module::Runtime was a drop in replacement for Class::MOP::load_class

0.63  2014-05-05 22:56:43
        - Make MooseX::MarkAsMethods a hard prerequisite (RT#94923)
        - Fix t/08helper.t with Data::Dumper >= 2.151 (RT#94599)

0.62  2014-01-06 12:32:27
        - Don't ship MYMETA.* files (RT#91808)
        - Stop using deprecated Class::MOP::load_module

0.61  2013-06-19 12:48:34
        - Fix test failure caused by hash randomisation in perl 5.17 (RT#82917)

0.60  2012-06-12 16:27:00
        - Add per_request_schema hook to PerRequestSchema trait and docs
        - Additional paranoia in types as it's possible for loading code
          to clobber $_

0.59  2011-11-01 11:20:46
        - update helper deps for new loader

0.58  2011-10-25 19:19:43
        - remove ->make_immutable from the PerRequestSchema trait

0.57  2011-10-22 16:01:45
        - add POD for PerRequestSchema trait

0.56  2011-10-22 15:34:59
        - add PerRequestSchema trait (t0m)

0.55  Fri Sep 16 08:55:53 UTC 2011
        - add DBD::SQLite to test_requires (RT#70357)

0.54  Fri Aug  5 11:29:00 EDT 2011
        - Changed t/08helper.t to skip if required features are not installed.
        - tweaked the Makefile.PL to support these changes

0.53  Wed Aug  3 03:45:07 UTC 2011
        - check loader args after connect_info against loader methods and remove
          them from connect_info

0.52  Thu Jul 28 22:07:38 UTC 2011
        - fix a bug in loader args reading after connect_info (skip structs)

0.51  Wed Jul 27 20:57:47 UTC 2011
        - warn when helper finds no tables
        - accept loader args after connect_info

0.50  Fri May 20 22:45:07 UTC 2011
        - change repository link to the new git repo
        - get t/05testapp.t to run on Win32

0.49  Wed May 11 06:03:50 UTC 2011
        - make sure storage_type class is loaded before ->isa check in
          Replicated trait (RT#65791)
        - fix regex stringification test for perl 5.14 (RT#68098)
        - update connect_info POD (RT#66010)

0.48  Thu Dec  9 21:08:33 UTC 2010
        - fix usage of Try::Tiny in helper

0.47  Wed Dec  8 22:21:06 UTC 2010
        - fix loader version check

0.46  Wed Dec  8 13:35:28 UTC 2010
        - make use_moose detection more robust

0.45  Wed Dec  8 12:05:58 UTC 2010
        - fix bug where non-result files were picked up for Moose check

0.44  Tue Dec  7 03:50:48 UTC 2010
        - do not upgrade non-Moose schemas to use_moose=1 (RT#60558)
        - added col_collision_map => 'column_%s' as default loader option
          (will take effect on release of loader 0.07003)

0.43  Sun Jul 25 01:00:34 UTC 2010
        - add dep for MooseX::NonMoose for the use_moose=1 option

0.42  Sat Jul 24 23:14:27 UTC 2010
        - add use_moose=1 to loader options by default for create=static

0.41  Wed Apr 28 08:42:13 EDT 2010
        - allow configuring ResultSets using the SchemaProxy trait

0.40  Wed Feb  3 23:12:16 EST 2010
        - fix Replicated trait deps

0.39  Mon Feb  1 10:08:51 EST 2010
        - better 'on_connect_do' parsing in Helper

0.38  Fri Jan 15 06:41:24 EST 2010
        - re-add the 'storage' proxy method

0.37  Thu Jan 14 10:20:43 EST 2010
        - support single replicant hashref from config

0.36  Thu Jan 14 09:21:10 EST 2010
        - upgrading a schema to use_namespaces=1 will also turn on
          components=InflateColumn::DateTime (the default.)

0.35  Sun Dec 27 04:50:53 EST 2009
        - forgot to edit Changes on last upload

0.34  Sun Dec 27 03:49:34 EST 2009
        - move schema proxying into SchemaProxy trait which is off by default

0.33  Sat Dec 26 08:04:49 EST 2009
        - fix test failure in 05testapp.t when env var not set (RT#53101)

0.32  Wed Dec 23 01:22:06 EST 2009
        - Move documentation on setting up authentication into its own section
          for clarity.
        - Other misc documentation enhancements to make the docs more clear
          about the important things to know.
        - Convert Authentication Plugin configuration example to new style.
        - Support Coderefs in loader options for helper
          (patch from Ozum Eldogan ozum@ozum.net)

0.31  Sun Oct 18 18:50:00 BST 2009
        - update for copyright info

0.30  Sun Oct 18 01:35:36 EDT 2009
        - change no sources error to a warning (with a way to turn it off)
        - write connect_info to Model without a create= option as well

0.29  Mon Sep  7 15:31:29 EDT 2009
        - support for coderef connect_info's


lib/Catalyst/Model/DBIC/Schema.pm
package Catalyst::Model::DBIC::Schema;

use Moose;
use mro 'c3';
extends 'Catalyst::Model';
with 'CatalystX::Component::Traits';

our $VERSION = '0.65';
$VERSION = eval $VERSION;

use namespace::autoclean;
use Carp::Clan '^Catalyst::Model::DBIC::Schema';
use Data::Dumper;
use DBIx::Class ();
use Module::Runtime qw/use_module/;

use Catalyst::Model::DBIC::Schema::Types
    qw/ConnectInfo SchemaClass Schema/;

use MooseX::Types::Moose qw/Str Bool/;
use MooseX::Types::LoadableClass qw/LoadableClass/;

=head1 NAME

Catalyst::Model::DBIC::Schema - DBIx::Class::Schema Model Class

=head1 SYNOPSIS

First, prepare your database schema using L<DBIx::Class>, see
L<Catalyst::Helper::Model::DBIC::Schema> for how to generate a
L<DBIx::Class::Schema> from your database using the Helper script, and
L<DBIx::Class::Schema::Loader::Base>.

A typical usage of the helper script would be:

    script/myapp_create.pl model FilmDB DBIC::Schema MyApp::Schema::FilmDB \
        create=static dbi:mysql:filmdb dbusername dbpass \
        quote_names=1

If you are unfamiliar with L<DBIx::Class>, see L<DBIx::Class::Manual::Intro>
first.

These examples assume that you already have a schema called
C<MyApp::Schema::FilmDB>, which defines some Result classes for tables in
C<MyApp::Schema::FilmDB::Result::Actor> and
C<MyApp::Schema::FilmDB::Result::Film>. Either created by the helper script (as
shown above) or manually.

The helper also creates a Model in C<lib/MyApp/Model/FilmDB.pm>, if you already
have a schema you can create just the Model using:

    script/myapp_create.pl model FilmDB DBIC::Schema MyApp::Schema::FilmDB
        dbi:mysql:filmdb dbusername dbpass

The connect_info is optional and will be hardcoded into the Model if provided.
It's better to configure it in your L<Catalyst> config file, which will also
override any hardcoded config, see L</connect_info> for examples.

Now you have a working Model which accesses your separate DBIC Schema. This can
be used/accessed in the normal Catalyst manner, via C<< $c->model() >>:

  my $db_model = $c->model('FilmDB');         # a Catalyst::Model
  my $dbic     = $c->model('FilmDB')->schema; # the actual DBIC object

There is also a shortcut, which returns a L<DBIx::Class::ResultSet> directly,
instead of a L<Catalyst::Model>:

  my $rs = $c->model('FilmDB::Actor');

See L<DBIx::Class::ResultSet> to find out more about which methods can be
called on ResultSets.

You can also define your own ResultSet methods to encapsulate the
database/business logic of your applications. These go into, for example,
C<lib/MyApp/Schema/FilmDB/ResultSet/Actor.pm>. The class must inherit from
L<DBIx::Class::ResultSet> and is automatically loaded.

Then call your methods like any other L<DBIx::Class::ResultSet> method:

    $c->model('FilmDB::Actor')->SAG_members

=head2 Some examples:

  # to access schema methods directly:
  $c->model('FilmDB')->schema->source(...);

  # to access the source object, resultset, and class:
  $c->model('FilmDB')->source(...);
  $c->model('FilmDB')->resultset(...);
  $c->model('FilmDB')->class(...);

  # For resultsets, there's an even quicker shortcut:
  $c->model('FilmDB::Actor')
  # is the same as $c->model('FilmDB')->resultset('Actor')

  # To get the composed schema for making new connections:
  my $newconn = $c->model('FilmDB')->composed_schema->connect(...);

  # Or the same thing via a convenience shortcut:
  my $newconn = $c->model('FilmDB')->connect(...);

  # or, if your schema works on different storage drivers:
  my $newconn = $c->model('FilmDB')->composed_schema->clone();
  $newconn->storage_type('::LDAP');
  $newconn->connection(...);

  # and again, a convenience shortcut
  my $newconn = $c->model('FilmDB')->clone();
  $newconn->storage_type('::LDAP');
  $newconn->connection(...);

To set up authentication, see L</"Setting up DBIC authentication"> below.

=head1 DESCRIPTION

This is a Catalyst Model for L<DBIx::Class::Schema>-based Models.  See
the documentation for L<Catalyst::Helper::Model::DBIC::Schema> for
information on generating these Models via Helper scripts.

When your Catalyst app starts up, a thin Model layer is created as an interface
to your DBIC Schema. It should be clearly noted that the model object returned
by C<< $c->model('FilmDB') >> is NOT itself a DBIC schema or resultset object,
but merely a wrapper proving L<methods|/METHODS> to access the underlying
schema.

In addition to this model class, a shortcut class is generated for each 
source in the schema, allowing easy and direct access to a resultset of the 
corresponding type. These generated classes are even thinner than the model 
class, providing no public methods but simply hooking into Catalyst's 
model() accessor via the 
L<ACCEPT_CONTEXT|Catalyst::Component/ACCEPT_CONTEXT> mechanism. The complete 
contents of each generated class is roughly equivalent to the following:

  package MyApp::Model::FilmDB::Actor
  sub ACCEPT_CONTEXT {
      my ($self, $c) = @_;
      $c->model('FilmDB')->resultset('Actor');
  }

In short, there are three techniques available for obtaining a DBIC 
resultset object: 

  # the long way
  my $rs = $c->model('FilmDB')->schema->resultset('Actor');

  # using the shortcut method on the model object
  my $rs = $c->model('FilmDB')->resultset('Actor');

  # using the generated class directly
  my $rs = $c->model('FilmDB::Actor');

In order to add methods to a DBIC resultset, you cannot simply add them to 
the source (row, table) definition class; you must define a separate custom 
resultset class. This is just a matter of making a
C<lib/MyApp/Schema/ResultSet/Actor.pm> class that inherits from
L<DBIx::Class::ResultSet>, if you are using
L<DBIx::Class::Schema/load_namespaces>, the default for helper script generated
schemas.

See L<DBIx::Class::Manual::Cookbook/"Predefined searches"> 
for information on definining your own L<DBIx::Class::ResultSet> classes for
use with L<DBIx::Class::Schema/load_classes>, the old default.

=head1 CONFIG PARAMETERS

=head2 schema_class

This is the classname of your L<DBIx::Class::Schema> Schema.  It needs
to be findable in C<@INC>, but it does not need to be inside the 
C<Catalyst::Model::> namespace.  This parameter is required.

=head2 connect_info

This is a hashref or arrayref of connection parameters, which are specific to
your C<storage_type> (see your storage type documentation for more details). If
you only need one parameter (e.g. the DSN), you can just pass a string.

This is not required if C<schema_class> already has connection information
defined inside itself (which isn't highly recommended, but can be done.)

For L<DBIx::Class::Storage::DBI>, which is the only supported
C<storage_type> in L<DBIx::Class> at the time of this writing, the
parameters are your dsn, username, password, and connect options hashref.

See L<DBIx::Class::Storage::DBI/connect_info> for a detailed explanation
of the arguments supported.

Examples:

  connect_info => {
    dsn => 'dbi:Pg:dbname=mypgdb',
    user => 'postgres',
    password => ''
  }

  connect_info => {
    dsn => 'dbi:SQLite:dbname=foo.db',
    on_connect_do => [
      'PRAGMA synchronous = OFF',
    ]
  }

  connect_info => {
    dsn => 'dbi:Pg:dbname=mypgdb',
    user => 'postgres',
    password => '',
    pg_enable_utf8 => 1,
    on_connect_do => [
      'some SQL statement',
      'another SQL statement',
    ],
  }

Or using L<Config::General>:

    <Model::FilmDB>
        schema_class   MyApp::Schema::FilmDB
        traits Caching
        <connect_info>
            dsn   dbi:Pg:dbname=mypgdb
            user   postgres
            password ""
            auto_savepoint 1
            quote_names 1
            on_connect_do   some SQL statement
            on_connect_do   another SQL statement
        </connect_info>
        user_defined_schema_accessor foo
    </Model::FilmDB>

or

    <Model::FilmDB>
        schema_class   MyApp::Schema::FilmDB
        connect_info   dbi:SQLite:dbname=foo.db
    </Model::FilmDB>

Or using L<YAML>:

  Model::MyDB:
      schema_class: MyDB
      traits: Caching
      connect_info:
          dsn: dbi:Oracle:mydb
          user: mtfnpy
          password: mypass
          LongReadLen: 1000000
          LongTruncOk: 1
          on_connect_call: 'datetime_setup'
          quote_names: 1

The old arrayref style with hashrefs for L<DBI> then L<DBIx::Class> options is also
supported:

  connect_info => [
    'dbi:Pg:dbname=mypgdb',
    'postgres',
    '',
    {
      pg_enable_utf8 => 1,
    },
    {
      auto_savepoint => 1,
      on_connect_do => [
        'some SQL statement',
        'another SQL statement',
      ],
    }
  ]

=head2 traits

Array of Traits to apply to the instance. Traits are L<Moose::Role>s.

They are relative to the C<< MyApp::TraitFor::Model::DBIC::Schema:: >>, then
the C<< Catalyst::TraitFor::Model::DBIC::Schema:: >> namespaces, unless
prefixed with C<+> in which case they are taken to be a fully qualified name.
E.g.:

    traits Caching
    traits +MyApp::TraitFor::Model::Foo

A new instance is created at application time, so any consumed required
attributes, coercions and modifiers will work.

Traits are applied at L<Catalyst::Component/COMPONENT> time using
L<CatalystX::Component::Traits>.

C<ref $self> will be an anon class if any traits are applied, C<<
$self->_original_class_name >> will be the original class.

When writing a Trait, interesting points to modify are C<BUILD>, L</setup> and
L</ACCEPT_CONTEXT>.

Traits that come with the distribution:

=over 4

=item L<Catalyst::TraitFor::Model::DBIC::Schema::Caching>

=item L<Catalyst::TraitFor::Model::DBIC::Schema::Replicated>

=item L<Catalyst::TraitFor::Model::DBIC::Schema::SchemaProxy>

=item L<Catalyst::TraitFor::Model::DBIC::Schema::PerRequestSchema>

=back

=head2 compose_namespaces

This model calls L<DBIx::Class::Schema/compose_namespace> by default to
install classes into the model namespaces. You can turn that off by
setting this attribute to false. Default is true.

=head2 install_model_shortcuts

If you don't want shortcut models so you can do e.g. C<< $c->model('DB::Book')
>> set this attribute to false, Default is true.

=head2 storage_type

Allows the use of a different C<storage_type> than what is set in your
C<schema_class> (which in turn defaults to C<::DBI> if not set in current
L<DBIx::Class>).  Completely optional, and probably unnecessary for most
people until other storage backends become available for L<DBIx::Class>.

=head1 ATTRIBUTES

The keys you pass in the model configuration are available as attributes.

Other attributes available:

=head2 connect_info

Your connect_info args normalized to hashref form (with dsn/user/password.) See
L<DBIx::Class::Storage::DBI/connect_info> for more info on the hashref form of
L</connect_info>.

=head2 model_name

The model name L<Catalyst> uses to resolve this model, the part after
C<::Model::> or C<::M::> in your class name. E.g. if your class name is
C<MyApp::Model::DB> the L</model_name> will be C<DB>.

=head2 _default_cursor_class

What to reset your L<DBIx::Class::Storage::DBI/cursor_class> to if a custom one
doesn't work out. Defaults to L<DBIx::Class::Storage::DBI::Cursor>.

=head1 ATTRIBUTES FROM L<MooseX::Traits::Pluggable>

=head2 _original_class_name

The class name of your model before any L</traits> are applied. E.g.
C<MyApp::Model::DB>.

=head2 _traits

Unresolved arrayref of traits passed in the config.

=head2 _resolved_traits

Traits you used resolved to full class names.

=head1 CONFIGURING YOUR SCHEMA AND RESULTSETS

See the documentation for
L<Catalyst::TraitFor::Model::DBIC::Schema::SchemaProxy> for instructions on how
to pass config values from your L<Catalyst> config to your
L<DBIx::Class::Schema> and/or L<DBIx::Class::ResultSet> classes.

=head1 METHODS

=head2 new

Instantiates the Model based on the above-documented ->config parameters.
The only required parameter is C<schema_class>.  C<connect_info> is
required in the case that C<schema_class> does not already have connection
information defined for it.

=head2 schema

Accessor which returns the connected schema being used by the this model.
There are direct shortcuts on the model class itself for
schema->resultset, schema->source, and schema->class.

=head2 composed_schema

Accessor which returns the composed schema, which has no connection info,
which was used in constructing the L</schema>. Useful for creating
new connections based on the same schema/model.  There are direct shortcuts
from the model object for composed_schema->clone and composed_schema->connect

If L</compose_namespaces> is not true, L</composed_schema> is equivalent to
C<< $model->schema_class->clone >>.

=head2 clone

Shortcut for ->composed_schema->clone

=head2 connect

Shortcut for ->composed_schema->connect

=head2 source

Shortcut for ->schema->source

=head2 class

Shortcut for ->schema->class

=head2 resultset

Shortcut for ->schema->resultset

=head2 txn_do

Shortcut for ->schema->txn_do

=head2 txn_scope_guard

Shortcut for ->schema->txn_scope_guard

=head2 storage

Provides an accessor for the connected schema's storage object.

See L<DBIx::Class::Storage> and L<DBIx::Class::Storage::DBI>.

=cut

has schema_class => (
    is => 'ro',
    isa => SchemaClass,
    required => 1
);

has compose_namespaces => (is => 'ro', isa => Bool, default => 1 );

has install_model_shortcuts => (is => 'ro', isa => Bool, default => 1 );

has storage_type => (is => 'rw', isa => Str);

has connect_info => (is => 'rw', isa => ConnectInfo, coerce => 1);

has model_name => (
    is => 'ro',
    isa => Str,
    required => 1,
    lazy_build => 1,
);

has _default_cursor_class => (
    is => 'ro',
    isa => LoadableClass,
    default => 'DBIx::Class::Storage::DBI::Cursor',
);

has schema => (is => 'rw', isa => Schema);

my $app_class;

before COMPONENT => sub {
    $app_class = ref $_[1] || $_[1];
};

sub app_class { $app_class }

sub BUILD {
    my ($self, $args) = @_;
    my $class = $self->_original_class_name;
    my $schema_class = $self->schema_class;

    if( !$self->connect_info ) {
        if($schema_class->storage && $schema_class->storage->connect_info) {
            $self->connect_info($schema_class->storage->connect_info);
        }
        else {
            die "Either ->config->{connect_info} must be defined for $class"
                  . " or $schema_class must have connect info defined on it."
		  . " Here's what we got:\n"
		  . Dumper($args);
        }
    }

    if (exists $self->connect_info->{cursor_class}) {
        eval { use_module($self->connect_info->{cursor_class}) }
            or croak "invalid connect_info: Cannot load your cursor_class"
        . " ".$self->connect_info->{cursor_class}.": $@";
    }

    $self->setup($args);

    my $is_installed = defined $self->composed_schema;

    if (not $is_installed) {
        $self->composed_schema($self->compose_namespaces ?
            $schema_class->compose_namespace($class)
            :
            $schema_class->clone
        );
    }

    $self->schema($self->composed_schema->clone)
        unless $self->schema;

    $self->schema->storage_type($self->storage_type)
        if $self->storage_type;

    $self->schema->connection($self->connect_info);

    if ((not $is_installed) && $self->install_model_shortcuts) {
        $self->_install_rs_models;
    }
}

sub clone { shift->composed_schema->clone(@_); }

sub connect { shift->composed_schema->connect(@_); }

# some proxy methods, see also SchemaProxy

sub resultset { shift->schema->resultset(@_); }

sub txn_do { shift->schema->txn_do(@_); }

sub txn_scope_guard { shift->schema->txn_scope_guard(@_); }

sub storage { shift->schema->storage(@_); }

=head2 setup

Called at C<BUILD> time before configuration, but after L</connect_info> is
set. To do something after configuuration use C<< after BUILD => >>.

Receives a hashref of args passed to C<BUILD>.

=cut

sub setup { 1 }

=head2 ACCEPT_CONTEXT

Point of extension for doing things at C<< $c->model >> time with context,
returns the model instance, see L<Catalyst::Manual::Intro/ACCEPT_CONTEXT> for
more information.

=cut

sub ACCEPT_CONTEXT { shift }

sub _install_rs_models {
    my $self  = shift;
    my $class = $self->_original_class_name;

    no strict 'refs';

    my @sources = $self->schema->sources;

    unless (@sources) {
        warn <<'EOF' unless $ENV{CMDS_NO_SOURCES};
******************************* WARNING ***************************************
* No sources found (did you forget to define your tables?)                    *
*                                                                             *
* To turn off this warning, set the CMDS_NO_SOURCES environment variable.     *
*******************************************************************************
EOF
    }

    foreach my $moniker (@sources) {
        my $classname = "${class}::$moniker";
        *{"${classname}::ACCEPT_CONTEXT"} = sub {
            shift;
            shift->model($self->model_name)->resultset($moniker);
        }
    }
}

sub _reset_cursor_class {
    my $self = shift;

    if ($self->storage->can('cursor_class')) {
	$self->storage->cursor_class($self->_default_cursor_class)
	    if $self->storage->cursor_class ne $self->_default_cursor_class;
    }
}

{
    my %COMPOSED_CACHE;

    sub composed_schema {
	my $self = shift;
	my $class = $self->_original_class_name;
	my $store = \$COMPOSED_CACHE{$class}{$self->schema_class};

	$$store = shift if @_;

	return $$store
    }
}

sub _build_model_name {
    my $self  = shift;
    my $class = $self->_original_class_name;
    (my $model_name = $class) =~ s/^[\w:]+::(?:Model|M):://;

    return $model_name;
}

__PACKAGE__->meta->make_immutable;

=head1 ENVIRONMENT

=over 4

=item CMDS_NO_SOURCES

Set this variable if you will be using schemas with no sources (Result classes)
to disable the warning. The warning is there because having no Result classes
is usually a mistake.

=back

=head1 Setting up DBIC authentication

You can set this up with 
L<Catalyst::Authentication::Store::DBIx::Class> in MyApp.pm:

  package MyApp;

  use Catalyst qw/... Authentication .../;

  ...

  __PACKAGE__->config('Plugin::Authentication' =>
                {
                    default_realm => 'members',
                    members => {
                        credential => {
                            class => 'Password',
                            password_field => 'password',
                            password_type => 'hashed'
                            password_hash_type => 'SHA-256'
                        },
                        store => {
                            class => 'DBIx::Class',
                            user_model => 'DB::User',
                            role_relation => 'roles',
                            role_field => 'rolename',
                        }
                    }
                });

=head1 METHOD PROXYING

The automatic proxying to the underlying L<DBIx::Class::Schema> has been
removed as of version C<0.34>, to enable this feature add C<SchemaProxy> to
L</traits>.

See L<Catalyst::TraitFor::Model::DBIC::Schema::SchemaProxy>.

=head1 SEE ALSO

General Catalyst Stuff:

L<Catalyst::Manual>, L<Catalyst::Test>, L<Catalyst::Request>,
L<Catalyst::Response>, L<Catalyst::Helper>, L<Catalyst>,

Stuff related to DBIC and this Model style:

L<DBIx::Class>, L<DBIx::Class::Schema>,
L<DBIx::Class::Schema::Loader>, L<Catalyst::Helper::Model::DBIC::Schema>,
L<CatalystX::Component::Traits>, L<MooseX::Traits::Pluggable>

Traits:

L<Catalyst::TraitFor::Model::DBIC::Schema::Caching>,
L<Catalyst::TraitFor::Model::DBIC::Schema::Replicated>,
L<Catalyst::TraitFor::Model::DBIC::Schema::SchemaProxy>,
L<Catalyst::TraitFor::Model::DBIC::Schema::PerRequestSchema>,
L<Catalyst::TraitFor::Model::DBIC::Schema::QueryLog>

=head1 AUTHOR

Brandon L Black C<blblack at gmail.com>

=head1 CONTRIBUTORS

caelum: Rafael Kitover C<rkitover at cpan.org>

dandv: Dan Dascalescu C<dandv at cpan.org>

bluefeet: Aran Deltac C<bluefeet@cpan.org>

t0m: Tomas Doran C<bobtfish@bobtfish.net>

osfameron: C<osfameron@cpan.org>

ozum: Ozum Eldogan C<ozum@ozum.net>

Pavel I. Shaydo C<zwon@trinitum.org>

SineSwiper: Brendan Byrd <byrd.b@insightcom.com>

=head1 COPYRIGHT

Copyright (c) 2006 - 2010
the Catalyst::Model::DBIC::Schema L</AUTHOR> and L</CONTRIBUTORS>
as listed above.

=head1 LICENSE

This program is free software. You can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
# vim:sts=4 sw=4 et tw=80:
